heartbeat/monitor/monitor.go

```diff
@@ -0,0 +1,200 @@
+			heartbeat.ErrInvalidTimeDuration, args.HideInactiveValidatorInterval, minDuration)
+	}
+
+	return nil
+}
+
+// GetHeartbeats returns the heartbeat status
+func (monitor *heartbeatV2Monitor) GetHeartbeats() []data.PubKeyHeartbeat {
+	monitor.numInstances = make(map[string]uint64, 0)
+
+	pids := monitor.cache.Keys()
+
+	heartbeatsV2 := make([]data.PubKeyHeartbeat, 0)
+	for idx := 0; idx < len(pids); idx++ {
+		pid := pids[idx]
+		peerId := core.PeerID(pid)
+		hb, ok := monitor.cache.Get(pid)
+		if !ok {
+			log.Debug("could not get data from cache for pid", "pid", peerId.Pretty())
+			continue
+		}
+
+		heartbeatData, err := monitor.parseMessage(peerId, hb)
+		if err != nil {
+			log.Debug("could not parse message for pid", "pid", peerId.Pretty(), "error", err.Error())
+			continue
+		}
+
+		heartbeatsV2 = append(heartbeatsV2, heartbeatData)
+	}
+
+	for idx := range heartbeatsV2 {
+		pk := heartbeatsV2[idx].PublicKey
+		heartbeatsV2[idx].NumInstances = monitor.numInstances[pk]
+	}
+
+	sort.Slice(heartbeatsV2, func(i, j int) bool {
+		return strings.Compare(heartbeatsV2[i].PublicKey, heartbeatsV2[j].PublicKey) < 0
+	})
+
+	return heartbeatsV2
+}
+
+func (monitor *heartbeatV2Monitor) parseMessage(pid core.PeerID, message interface{}) (data.PubKeyHeartbeat, error) {
+	pubKeyHeartbeat := data.PubKeyHeartbeat{}
+
+	heartbeatV2, ok := message.(heartbeat.HeartbeatV2)
+	if !ok {
+		return pubKeyHeartbeat, process.ErrWrongTypeAssertion
+	}
+
+	payload := heartbeat.Payload{}
+	err := monitor.marshaller.Unmarshal(&payload, heartbeatV2.Payload)
+	if err != nil {
+		return pubKeyHeartbeat, err
+	}
+
+	peerInfo := monitor.peerShardMapper.GetPeerInfo(pid)
+
+	crtTime := time.Now()
+	messageAge := monitor.getMessageAge(crtTime, payload.Timestamp)
+	stringType := string(rune(peerInfo.PeerType))
+	if monitor.shouldSkipMessage(messageAge, stringType) {
+		return pubKeyHeartbeat, fmt.Errorf("validator should be skipped")
+	}
+
+	pk := monitor.pubKeyConverter.Encode(peerInfo.PkBytes)
+	monitor.numInstances[pk]++
+
+	pubKeyHeartbeat = data.PubKeyHeartbeat{
+		PublicKey:       pk,
+		TimeStamp:       crtTime,
+		IsActive:        monitor.isActive(messageAge),
+		ReceivedShardID: monitor.shardId,
+		ComputedShardID: peerInfo.ShardID,
+		VersionNumber:   heartbeatV2.GetVersionNumber(),
+		NodeDisplayName: heartbeatV2.GetNodeDisplayName(),
+		Identity:        heartbeatV2.GetIdentity(),
+		PeerType:        stringType,
+		Nonce:           heartbeatV2.GetNonce(),
+		PeerSubType:     heartbeatV2.GetPeerSubType(),
+		PidString:       pid.Pretty(),
+	}
+
+	return pubKeyHeartbeat, nil
+}
+
+func (monitor *heartbeatV2Monitor) getMessageAge(crtTime time.Time, messageTimestamp int64) time.Duration {
+	messageTime := time.Unix(messageTimestamp, 0)
+	msgAge := crtTime.Sub(messageTime)
+	return msgAge
+}
+
+func (monitor *heartbeatV2Monitor) isActive(messageAge time.Duration) bool {
+	if messageAge < 0 {
```

Comment:
let's not test this with 0 as the current's node RTC might be slightly behind others. Let's use a const duration such as -10 seconds

Reply:
Used the same approach as in old heartbeat implementation, with a max between 0 and current value.